### ISSUE TITLE
`redundant_closure_for_method_calls` suggesting a private type

### ISSUE THREAD

**Issue by lyssieth** [rust-lang/rust-clippy#10095]
### Summary

I noticed this bug while working on a project using https://github.com/lyssieth/htp/tree/lyssieth-time-support, specifically that  it was trying to use a private module from that branch of `htp` 

### Lint Name

redundant_closure_for_method_calls

### Reproducer

I tried this code:

```rust
// `expire` here is an Option<String> generally containing text like: `in 1 hour`, `in 50 min`, etc
let expire = match expire {
    Some(expire) => Some(htp::parse(&expire, OffsetDateTime::now_utc())?),
    None => None,
};

// this causes the false positive, since `unified` is an internal module for `htp`
// Converting the internal `htp::unified::DateTime` to a usable `time::OffsetDateTime` via the fn `as_time`
let expire = expire.and_then(|dt|dt.as_time());
// after application:
let expire = expire.and_then(htp::unified::DateTime::as_time);
// and this then causes an error.
```

I saw this happen:

```
warning: redundant closure
  --> src/main.rs:63:46
   |
63 |                 let expire = expire.and_then(|dt| dt.as_time());
   |                                              ^^^^^^^^^^^^^^^^^ help: replace the closure with the method itself: `htp::unified::DateTime::as_time`
   |
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#redundant_closure_for_method_calls
note: the lint level is defined here
  --> src/main.rs:1:9
   |
1  | #![warn(clippy::pedantic)]
   |         ^^^^^^^^^^^^^^^^
   = note: `#[warn(clippy::redundant_closure_for_method_calls)]` implied by `#[warn(clippy::pedantic)]`
```

I expected to see this happen:

It to not trigger, as if the lint is applied, it causes an error instead. `htp::unified` is a private module of `htp`.

### Version

```text
rustc 1.66.0 (69f9c33d7 2022-12-12)
binary: rustc
commit-hash: 69f9c33d71c871fc16ac445211281c6e7a340943
commit-date: 2022-12-12
host: x86_64-unknown-linux-gnu
release: 1.66.0
LLVM version: 15.0.2
```

### Additional Labels

@rustbot label +I-suggestion-causes-error

**Comment by feniljain** [comment]
I think the MRE should look something like this:

```rust
#![warn(clippy::pedantic)]

mod a {
    mod b {
        pub struct C(pub i32);

        impl C {
            pub fn get_val(&self) -> Option<i32> {
                Some(self.0)
            }
        }
    }

    pub fn get_c() -> b::C {
        b::C(0)
    }
}

fn main() {
	    let opt = Some(a::get_c());
		let _a = opt.and_then(|x| x.get_val());
}
```

Tho unfortunately it does not give this suggestion for me, maybe I am missing something, any ideas?

**Comment by lyssieth** [comment]
I'll try to pull out a MRE from the project sometime later today or tomorrow, if one can't be figured out beforehand.

**Comment by Holzhaus** [comment]
Looks like this was fixed in 1.83.0, at least i can't reproduce it anymore.
